src/psyclone/psyad/tl2ad.py

```diff
@@ -98,9 +155,407 @@ def generate_adjoint(tl_psyir):
 
     # Transform from TL to AD
     logger.debug("Transformation from TL to AD should be done now.")
-    ad_psyir = tl_psyir
+    ad_psyir = tl_psyir.copy()
+
+    # We permit the input code to be a single Program or Subroutine
+    container = _find_container(ad_psyir)
+    if container:
+        # Re-name the Container for the adjoint code
+        container.name = container.name + name_suffix
+
+    routines = ad_psyir.walk(Routine)
+
+    if not routines:
+        raise InternalError("The supplied PSyIR does not contain any "
+                            "routines.")
+
+    if len(routines) != 1:
+        raise NotImplementedError(
+            "The supplied Fortran must contain one and only one routine "
+            "but found: {0}".format([sub.name for sub in routines]))
+    routine = routines[0]
+
+    # We need to re-name the kernel routine. Have to take care in case we've
+    # been supplied with a bare program/subroutine rather than a subroutine
+    # within a module.
+    if container:
+        kernel_sym = container.symbol_table.lookup(routine.name)
+        adj_kernel_name = routine.name + name_suffix
+        # A symbol's name is immutable so create a new RoutineSymbol
+        adj_kernel_sym = container.symbol_table.new_symbol(
+            adj_kernel_name, symbol_type=RoutineSymbol,
+            visibility=kernel_sym.visibility)
+        container.symbol_table.remove(kernel_sym)
+        routine.name = adj_kernel_sym.name
+    else:
+        routine.name = routine.name + name_suffix
 
     return ad_psyir
 
 
-__all__ = ["generate_adjoint_str", "generate_adjoint"]
+def generate_adjoint_test(tl_psyir, ad_psyir):
```

Comment:
In general I think it would be useful to add logging info as we go along. So perhaps output the new names being used the arguments and their names etc.

Reply:
Ooh, it feels weird doing that. (Like going back to print statements for debugging.) It's nice to see the output though (once I'd figured out how to see it).